crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,261 @@
+    }
+
+    /// Reserves capacity for at least `additional` bytes more than the current length.
+    #[inline]
+    pub fn reserve(&mut self, additional: usize) {
+        // `self.len()` counts the number of bytes excluding the terminating nul, but the
+        // size passed to `sys::furi_string_reserve` needs to include the nul terminator.
+        unsafe { sys::furi_string_reserve(self.0, self.len() + additional + 1) };
+    }
+
+    /// Appends the given [`char`] to the end of this `String`.
+    #[inline]
+    pub fn push(&mut self, ch: char) {
+        match ch.len_utf8() {
+            1 => unsafe { sys::furi_string_push_back(self.0, ch as c_char) },
+            _ => unsafe { sys::furi_string_utf8_push(self.0, ch as u32) },
+        }
+    }
+
+    /// Returns a byte slice of this `String`'s contents.
+    ///
+    /// The returned slice will **not** contain the trailing nul terminator that the
+    /// underlying C string has.
+    #[inline]
+    #[must_use]
+    pub fn to_bytes(&self) -> &[u8] {
+        self.as_c_str().to_bytes()
+    }
+
+    /// Returns a byte slice of this `String`'s contents with the trailing nul byte.
+    ///
+    /// This function is the equivalent of [`String::to_bytes`] except that it will retain
+    /// the trailing nul terminator instead of chopping it off.
+    #[inline]
+    #[must_use]
+    pub fn to_bytes_with_nul(&self) -> &[u8] {
+        self.as_c_str().to_bytes_with_nul()
+    }
+
+    /// Shortens this `String` to the specified length.
+    ///
+    /// If `new_len` is greater than the string's current length, this has no effect.
+    #[inline]
+    pub fn truncate(&mut self, new_len: usize) {
+        unsafe { sys::furi_string_left(self.0, new_len) };
+    }
+
+    /// Returns the length of this `String`.
+    ///
+    /// This length is in bytes, not [`char`]s or graphemes. In other words, it might not
+    /// be what a human considers the length of the string.
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        unsafe { sys::furi_string_size(self.0) }
+    }
+
+    /// Returns `true` if this `String` has a length of zero, and `false` otherwise.
+    #[inline]
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        unsafe { sys::furi_string_empty(self.0) }
+    }
+
+    /// Splits the string into two at the given byte index.
+    ///
+    /// Returns a newly allocated `String`. `self` contains bytes `[0, at)`, and the
+    /// returned `String` contains bytes `[at, len)`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `at` is beyond the last byte of the string.
+    #[inline]
+    #[must_use = "use `.truncate()` if you don't need the other half"]
+    pub fn split_off(&mut self, at: usize) -> String {
+        // SAFETY: Trimming the beginning of a C string results in a valid C string, as
+        // long as the nul byte is not trimmed.
+        assert!(at <= self.len());
```

Comment:
[`String::split_off`](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.split_off) similarly panics if this condition is not met, as well as if `at` is not at a `char` boundary in the UTF-8 encoded string. For now I've dropped the UTF-8 panic (given that `FuriString` doesn't try to preserve UTF-8 encoding) but kept the OOB panic as-is (there are plenty of similar panics we can't avoid due to the Flipper Zero SDK not exposing error handling for these methods).

If we want to avoid the panic then we could instead make this `try_split_off(&mut self, at: usize) -> Result<String, SplitOffError>`.

Reply:
That seems fine to me. 👍